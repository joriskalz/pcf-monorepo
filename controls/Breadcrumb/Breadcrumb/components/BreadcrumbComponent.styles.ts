// This wrapperstyle is required, so that,
// the breadcrumb don't shrink when its inside
// flexbox, which happens in studio
export const wrapperStyle = {
    bottom: 0,
    left: 0,
    right: 0,
    top: 0,
    overflowX: 'hidden',
    overflowY: 'hidden',
    position: 'absolute',
    width: '100%',
    height: '100%',
} as React.CSSProperties;
