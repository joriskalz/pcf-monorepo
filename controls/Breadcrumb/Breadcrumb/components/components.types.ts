export interface ICustomBreadcrumbItem {
    id: string;
    text: string;
    key: string;
    clickable?: boolean;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
}

export interface IBreadcrumbProps {
    width?: number;
    height?: number;
    items: ICustomBreadcrumbItem[];
    onSelected: (item: ICustomBreadcrumbItem) => void;
    themeJSON?: string;
    ariaLabel?: string;
    setFocus?: string;
    tabIndex?: number;
    maxDisplayedItems?: number;
    overflowIndex?: number;
}
