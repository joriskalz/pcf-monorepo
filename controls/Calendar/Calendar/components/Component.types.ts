export interface ICalendarProps {
    width?: number;
    height?: number;
    themeJSON?: string;
    setFocus?: string;
    tabIndex?: number;
    selectedDateValue: Date | undefined;
    minDate: Date | undefined;
    maxDate: Date | undefined;
    ariaLabel: string;
    showGoToToday?: boolean;
    monthPickerVisible?: boolean;
    dayPickerVisible?: boolean;
    highlightSelectedMonth?: boolean;
    highlightCurrentMonth?: boolean;
    showSixWeeksByDefault?: boolean;
    showWeekNumbers: boolean;
    firstDayOfWeek: string;
    isDisabled: boolean;
    language: string;
    // eslint-disable-next-line
    onSelected: (selectedDateValue: Date) => void;
    backgroundColor?: string;
}
