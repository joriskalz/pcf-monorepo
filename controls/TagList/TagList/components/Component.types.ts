export interface TagListItem {
    id: string;
    enabled?: boolean;
    name: string;
    iconName?: string;
    iconColor?: string;
    backgroundColor?: string;
    borderColor?: string;
    textColor?: string;
    iconOnly?: boolean;
    visible?: boolean;
    parentItemKey?: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
    key: string;
}

export interface CanvasTagListProps {
    width?: number;
    height?: number;
    items: TagListItem[];
    disabled?: boolean;
    tabIndex?: number;
    ariaLabel?: string;
    themeJSON?: string;
    itemMaxWidth?: number;
    maxHeight?: number;
    itemHeight?: number;
    fontSize?: number;
    borderRadius?: number;
    justify?: 'center' | 'left' | 'right';
    onResize?: (width: number, height: number) => void;
}
