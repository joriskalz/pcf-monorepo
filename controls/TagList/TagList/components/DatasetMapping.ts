import { TagListItem } from './Component.types';
import { ItemColumns } from '../ManifestConstants';

export function getTagItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): TagListItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            name: record.getValue(ItemColumns.DisplayName) as string,
            enabled: record.getValue(ItemColumns.Enabled) as boolean,
            iconName: record.getValue(ItemColumns.IconName) as string,
            iconColor: record.getValue(ItemColumns.IconColor) as string,
            textColor: record.getValue(ItemColumns.TextColor) as string,
            borderColor: record.getValue(ItemColumns.BorderColor) as string,
            backgroundColor: record.getValue(ItemColumns.BackgroundColor) as string,
            iconOnly: record.getValue(ItemColumns.IconOnly) as boolean,
            visible: record.getValue(ItemColumns.Visible) as boolean,
            data: record,
        } as TagListItem;
    });
}

function getDummyAction(key: string): TagListItem {
    return {
        id: key,
        key: key,
        name: 'Item ' + key,
        iconName: 'Unknown',
    } as TagListItem;
}
