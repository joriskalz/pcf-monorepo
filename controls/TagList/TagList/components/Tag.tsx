import { useId } from '@fluentui/react-hooks';
import {
    FontIcon,
    classNamesFunction,
    styled,
    ITagItemProps,
    ITagItemStyleProps,
    ITagItemStyles,
} from '@fluentui/react';
import { getStyles } from './TagItem.styles';
import * as React from 'react';

const getClassNames = classNamesFunction<ITagItemStyleProps, ITagItemStyles>();
export interface TagProps extends ITagItemProps {
    fontSize?: number;
    height?: number;
    iconName?: string;
    iconColor?: string;
    textColor?: string;
    backgroundColor?: string;
    borderColor?: string;
    borderRadius?: number;
}

// This is simlar to the TagItem in Fluent
export const TagComponent = (props: TagProps): JSX.Element => {
    const {
        theme,
        styles,
        iconColor,
        textColor,
        backgroundColor,
        borderColor,
        iconName,
        children,
        index,
        title = typeof props.children === 'string' ? props.children : props.item.name,
    } = props;

    const tagStyle = React.useMemo(() => {
        return {
            color: textColor,
            backgroundColor: backgroundColor,
            borderColor: borderColor,
            borderWidth: borderColor ? 1 : undefined,
            borderStyle: borderColor ? 'solid' : undefined,
        } as React.CSSProperties;
    }, [textColor, backgroundColor, borderColor]);

    const iconStyle = React.useMemo(() => {
        return {
            color: iconColor,
        } as React.CSSProperties;
    }, [iconColor]);

    const itemId = useId();
    const classNames = getClassNames(styles, {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        theme: theme!,
        disabled: true,
        selected: false,
    });

    return (
        <div className={classNames.root} style={tagStyle} role={'listitem'} aria-label={title} key={index}>
            {iconName && <FontIcon role="img" iconName={iconName} style={iconStyle} />}
            <span className={classNames.text} title={title} id={`${itemId}-text`}>
                {children}
            </span>
        </div>
    );
};

export const Tag = React.memo(
    styled<TagProps, ITagItemStyleProps, ITagItemStyles>(TagComponent, getStyles, undefined, {
        scope: 'TagItem',
    }),
);
