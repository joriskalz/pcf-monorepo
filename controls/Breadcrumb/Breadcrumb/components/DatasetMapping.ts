import { IBreadcrumbItem } from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';
import { ICustomBreadcrumbItem } from './components.types';

export function getItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): ICustomBreadcrumbItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            text: record.getFormattedValue(ItemColumns.DisplayName),
            clickable: record.getValue(ItemColumns.Clickable) as boolean,
            data: record,
        } as ICustomBreadcrumbItem;
    });
}

export function getBreadcrumbItems(
    items: ICustomBreadcrumbItem[],
    onClick: (ev?: React.MouseEvent<HTMLElement>, item?: ICustomBreadcrumbItem) => void,
): IBreadcrumbItem[] {
    return items.map((item) =>
        item.clickable === false
            ? {
                  id: item.id,
                  key: item.key,
                  text: item.text,
              }
            : {
                  id: item.id,
                  key: item.key,
                  text: item.text,
                  onClick: onClick,
              },
    ) as IBreadcrumbItem[];
}

function getDummyAction(key: string): ICustomBreadcrumbItem {
    return {
        id: key,
        key: key,
        text: 'Item ' + key,
        clickable: true,
    } as ICustomBreadcrumbItem;
}
