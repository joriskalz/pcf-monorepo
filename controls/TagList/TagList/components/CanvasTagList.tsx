import {
    ITagItemStyleProps,
    ITagItemStyles,
    ThemeProvider,
    IStyleFunctionOrObject,
    createTheme,
    IPartialTheme,
} from '@fluentui/react';
import * as React from 'react';
import { Tag } from './Tag';
import { TagListItem, CanvasTagListProps } from './Component.types';
import useResizeObserver from '@react-hook/resize-observer';

export const CanvasTagList = React.memo((props: CanvasTagListProps) => {
    const {
        items,
        width,
        height,
        maxHeight,
        itemMaxWidth,
        disabled,
        tabIndex,
        themeJSON,
        justify,
        onResize,
        itemHeight,
        fontSize,
        borderRadius,
    } = props;

    const containerStyle: React.CSSProperties = React.useMemo(() => {
        return {
            maxWidth: width ? width : '100%',
            position: 'relative',
            display: 'flex',
            maxHeight: maxHeight || height,
            overflowY: 'auto',
            flexWrap: 'wrap',
            alignContent: 'flex-start',
        };
    }, [height, maxHeight, width]);

    const tagStyles: IStyleFunctionOrObject<ITagItemStyleProps, ITagItemStyles> = React.useMemo(() => {
        return {
            root: {
                maxWidth: itemMaxWidth || '100%',
                minWidth: undefined,
                justifyContent: justify ?? 'left',
                borderRadius: borderRadius ?? undefined,
                lineHeight: itemHeight ?? undefined,
                height: itemHeight ?? undefined,
                fontSize: fontSize ?? undefined,
            },
            text: {
                minWidth: undefined,
            },
        };
    }, [itemMaxWidth, justify, borderRadius, itemHeight, fontSize]);

    const theme = React.useMemo(() => {
        try {
            return themeJSON ? createTheme(JSON.parse(themeJSON) as IPartialTheme) : undefined;
        } catch (ex) {
            /* istanbul ignore next */
            console.error('Cannot parse theme', ex);
        }
    }, [themeJSON]);

    const target = React.useRef<HTMLElement>(null);
    useResizeObserver(target, (entry) => {
        if (onResize) onResize(entry.contentRect.width, entry.contentRect.height);
    });

    return (
        <>
            <ThemeProvider theme={theme} applyTo="none" style={containerStyle} ref={target}>
                {items &&
                    items.map((item: TagListItem, index: number) => (
                        <Tag
                            key={item.id}
                            tabIndex={disabled ? -1 : tabIndex}
                            styles={tagStyles}
                            index={index}
                            item={item}
                            textColor={item.textColor}
                            borderColor={item.borderColor}
                            backgroundColor={item.backgroundColor}
                            iconName={item.iconName}
                            iconColor={item.iconColor}
                            disabled={disabled === true || item.enabled === false}
                            aria-setsize={items.length}
                            maxLength={300}
                        >
                            {item.name}
                        </Tag>
                    ))}
            </ThemeProvider>
        </>
    );
});
CanvasTagList.displayName = 'CanvasTagList';
