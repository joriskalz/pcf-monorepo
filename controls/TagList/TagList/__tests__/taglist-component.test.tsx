import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { TagListItem, CanvasTagListProps } from '../components/Component.types';
import { CanvasTagList } from '../components/CanvasTagList';
import { Tag } from '../components/Tag';

const mutationObserverMock = jest.fn<MutationObserver, [MutationCallback]>().mockImplementation(() => {
    return {
        observe: jest.fn(),
        disconnect: jest.fn(),
        takeRecords: jest.fn(),
    };
});

const defaultProps = {
    items: [
        {
            id: '1',
            key: '1',
            name: 'tag1',
        },
    ] as TagListItem[],
    onSelected: jest.fn(),
} as CanvasTagListProps;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
(global as any).MutationObserver = mutationObserverMock;

describe('TagList', () => {
    beforeEach(() => jest.clearAllMocks());

    it('tag renders correctly with no style', async () => {
        expect.assertions(1);
        const component = renderer.create(<Tag index={1} item={defaultProps.items[0]} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('taglist renders correctly with no style', async () => {
        expect.assertions(1);
        const component = renderer.create(<CanvasTagList {...defaultProps} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
