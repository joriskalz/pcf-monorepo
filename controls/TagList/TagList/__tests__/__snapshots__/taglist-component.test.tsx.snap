// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagList tag renders correctly with no style 1`] = `
<div
  aria-label="tag1"
  className=
      ms-TagItem
      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        background: #f3f2f1;
        border-radius: 2px;
        box-sizing: content-box;
        color: #323130;
        cursor: default;
        display: flex;
        flex-shrink: 1;
        flex-wrap: nowrap;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        height: 26px;
        line-height: 26px;
        margin-bottom: 2px;
        margin-left: 2px;
        margin-right: 2px;
        margin-top: 2px;
        max-width: 300px;
        min-width: 0px;
        outline: transparent;
        position: relative;
        user-select: none;
      }
      &::-moz-focus-inner {
        border: 0;
      }
      .ms-Fabric--isFocusVisible &:focus:after {
        border: 1px solid #ffffff;
        bottom: 1px;
        content: "";
        left: 1px;
        outline: 1px solid #605e5c;
        position: absolute;
        right: 1px;
        top: 1px;
        z-index: 1;
      }
      &:hover {
        background: #f3f2f1;
      }
      @media screen and (-ms-high-contrast: active), (forced-colors: active){& {
        border-color: GrayText;
        border: 1px solid WindowText;
      }
  role="listitem"
  style={
    Object {
      "backgroundColor": undefined,
      "borderColor": undefined,
      "borderStyle": undefined,
      "borderWidth": undefined,
      "color": undefined,
    }
  }
>
  <span
    className=
        ms-TagItem-text
        {
          margin-bottom: 0;
          margin-left: 8px;
          margin-right: 8px;
          margin-top: 0;
          min-width: 30px;
          overflow: hidden;
          text-overflow: ellipsis;
          white-space: nowrap;
        }
        @media screen and (-ms-high-contrast: active), (forced-colors: active){& {
          color: GrayText;
        }
    id="id__0-text"
    title="tag1"
  />
</div>
`;

exports[`TagList taglist renders correctly with no style 1`] = `
<div
  className=""
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexWrap": "wrap",
      "maxHeight": undefined,
      "maxWidth": "100%",
      "overflowY": "auto",
      "position": "relative",
    }
  }
>
  <div
    aria-label="tag1"
    className=
        ms-TagItem
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          background: #f3f2f1;
          border-radius: 2px;
          box-sizing: content-box;
          color: #323130;
          cursor: default;
          display: flex;
          flex-shrink: 1;
          flex-wrap: nowrap;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: 26px;
          justify-content: left;
          line-height: 26px;
          margin-bottom: 2px;
          margin-left: 2px;
          margin-right: 2px;
          margin-top: 2px;
          max-width: 100%;
          min-width: 0px;
          outline: transparent;
          position: relative;
          user-select: none;
        }
        &::-moz-focus-inner {
          border: 0;
        }
        .ms-Fabric--isFocusVisible &:focus:after {
          border: 1px solid #ffffff;
          bottom: 1px;
          content: "";
          left: 1px;
          outline: 1px solid #605e5c;
          position: absolute;
          right: 1px;
          top: 1px;
          z-index: 1;
        }
        &:hover {
          background: #f3f2f1;
        }
        @media screen and (-ms-high-contrast: active), (forced-colors: active){& {
          border-color: GrayText;
          border: 1px solid WindowText;
        }
    role="listitem"
    style={
      Object {
        "backgroundColor": undefined,
        "borderColor": undefined,
        "borderStyle": undefined,
        "borderWidth": undefined,
        "color": undefined,
      }
    }
  >
    <span
      className=
          ms-TagItem-text
          {
            margin-bottom: 0;
            margin-left: 8px;
            margin-right: 8px;
            margin-top: 0;
            min-width: 30px;
            overflow: hidden;
            text-overflow: ellipsis;
            white-space: nowrap;
          }
          @media screen and (-ms-high-contrast: active), (forced-colors: active){& {
            color: GrayText;
          }
      id="id__2-text"
      title="tag1"
    >
      tag1
    </span>
  </div>
</div>
`;
